File: src/algo.js
/**
 * Hospital routing algorithm — loads hospital data from CSV, scores per data_explanation.
 * getOptimalHospital(nlp_extracted_data, current_lat, current_lon) only; dataset path defined below.
 * For real ETAs set REACT_APP_GMAPS_API_KEY; otherwise uses CSV column ambulance_travel_time_min_sim.
 */

// Which dataset to load. For the browser, this path is fetched from the app origin (place CSV in public/).
const HOSPITAL_DATASET_PATH = "austin_hospitals_demo - austin_hospitals_demo.csv";

// Overlay when sendPatient() is called: bed deltas and specialist_patients increments per hospital.
const sentPatientOverlay = {}; // hospital_name -> { edBedsDelta, icuBedsDelta, specialistPatientsDelta: { Cardiology, Trauma, Neurology } }

const SPECIALTY_NAMES = ["Cardiology", "Trauma", "Neurology"];

/** Map NLP required_specialty to CSV specialty name for load calculation. */
function nlpSpecialtyToColumn(specialty) {
  const s = (specialty || "General").toLowerCase();
  if (s.includes("cardiac") || s.includes("stemi") || s.includes("heart")) return "Cardiology";
  if (s.includes("trauma")) return "Trauma";
  if (s.includes("stroke") || s.includes("neuro")) return "Neurology";
  return null;
}

function getRequiredSpecialtiesFromNlp(nlpData) {
  const specs = nlpData?.required_specialties;
  if (Array.isArray(specs)) return specs.map((s) => nlpSpecialtyToColumn(s)).filter(Boolean);
  const one = nlpSpecialtyToColumn(nlpData?.required_specialty);
  return one ? [one] : [];
}

/** Specialist load = (csv specialist_patients + overlay) / specialists. Null if no specialists. */
function getSpecialistLoad(row, specialty, hospitalName) {
  if (!specialty || !SPECIALTY_NAMES.includes(specialty)) return null;
  const colDoc = `specialists_${specialty}`;
  const colPat = `specialist_patients_${specialty}`;
  const numDoctors = num(row[colDoc]);
  if (numDoctors <= 0) return null;
  const csvPatients = num(row[colPat]);
  const overlay = sentPatientOverlay[hospitalName]?.specialistPatientsDelta ?? {};
  const patients = Math.max(0, csvPatients + (overlay[specialty] ?? 0));
  return patients / numDoctors;
}

function hasSpecialistFor(row, specialty, hospitalName) {
  return getSpecialistLoad(row, specialty, hospitalName) !== null;
}

let cachedHospitals = null;

function parseCsv(text) {
  const lines = text.trim().split(/\r?\n/);
  if (lines.length < 2) return [];
  const header = lines[0].split(",").map((s) => s.trim());
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const vals = lines[i].split(",").map((s) => s.trim());
    const row = {};
    header.forEach((h, j) => (row[h] = vals[j] ?? ""));
    rows.push(row);
  }
  return rows;
}

async function loadHospitalsFromDataset() {
  if (cachedHospitals) return cachedHospitals;
  const base = (typeof process !== "undefined" && process.env?.PUBLIC_URL) ? process.env.PUBLIC_URL : "";
  const url = (base + "/" + HOSPITAL_DATASET_PATH).replace(/\/+/g, "/");
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to load dataset: ${url}`);
  const text = await res.text();
  const rows = parseCsv(text);
  cachedHospitals = rows;
  return rows;
}

function hasTraumaCapability(traumaLevel) {
  if (!traumaLevel || traumaLevel === "N/A") return false;
  return /^I\b|^II\b|^III|^IV/i.test(traumaLevel);
}

function hasStrokeCapability(strokeLevel) {
  if (!strokeLevel) return false;
  const u = strokeLevel.toLowerCase();
  if (u.includes("none") && !u.includes("capable")) return false;
  return u.includes("primary") || u.includes("comprehensive") || u.includes("capable");
}

function hasCardiacCapability(cardiacCathLab) {
  return String(cardiacCathLab || "").toLowerCase() === "yes";
}

function hasPediatricCapability(pediatricSpecialty) {
  if (!pediatricSpecialty) return false;
  const u = pediatricSpecialty.toLowerCase();
  return u.includes("yes") || u.includes("limited") || u.includes("nicu");
}

function specialtyRequiresCapability(specialty) {
  const s = (specialty || "General").toLowerCase();
  if (s.includes("trauma")) return "trauma";
  if (s.includes("stroke")) return "stroke";
  if (s.includes("cardiac") || s.includes("stemi") || s.includes("heart")) return "cardiac";
  if (s.includes("pediatric") || s.includes("peds")) return "pediatric";
  return null;
}

function hospitalHasRequiredCapability(row, requiredCap) {
  if (!requiredCap) return true;
  switch (requiredCap) {
    case "trauma":
      return hasTraumaCapability(row.trauma_level);
    case "stroke":
      return hasStrokeCapability(row.stroke_center_level);
    case "cardiac":
      return hasCardiacCapability(row.cardiac_cath_lab);
    case "pediatric":
      return hasPediatricCapability(row.pediatric_specialty);
    default:
      return true;
  }
}

function applyHardFilters(rows, nlpData) {
  const acuity = Number(nlpData?.acuity_level) || 3;
  const requiredCap = specialtyRequiresCapability(nlpData?.required_specialty);
  const isCritical = acuity <= 2;

  return rows.filter((row) => {
    if (String(row.ed_diversion_sim || "").toLowerCase() === "yes") return false;
    if (!hospitalHasRequiredCapability(row, requiredCap)) return false;
    if (isCritical) {
      const icuAvail = num(row.available_icu_beds_sim) + (sentPatientOverlay[row.hospital_name]?.icuBedsDelta ?? 0);
      if (icuAvail <= 0) return false;
    }
    return true;
  });
}

function num(v) {
  const n = parseFloat(v);
  return Number.isFinite(n) ? n : 0;
}

// Scoring: lower is better. Includes specialist load (patients/doctors) for required_specialty.
function scoreHospital(row, etaMinutes, nlpData) {
  const acuity = Number(nlpData?.acuity_level) || 3;
  const travelWeight = 1.0;
  const waitWeight = 0.5;
  const capacityWeight = 2.0;
  const staffWeight = 0.3;
  const specialistLoadWeight = 20; // higher load = higher score (worse)
  const noSpecialistPenalty = 500;

  const erWait = num(row.er_wait_min_sim);
  const edBedsRaw = num(row.available_ed_beds_sim);
  const overlay = sentPatientOverlay[row.hospital_name] || {};
  const availableEdBeds = Math.max(0, edBedsRaw + (overlay.edBedsDelta ?? 0));
  const physicians = num(row.on_call_ed_physicians_sim);
  const icuAvail = Math.max(0, num(row.available_icu_beds_sim) + (overlay.icuBedsDelta ?? 0));

  let score = travelWeight * etaMinutes + waitWeight * erWait;
  score -= capacityWeight * availableEdBeds;
  score -= staffWeight * physicians;
  if (acuity <= 2 && icuAvail <= 1) score += 50;

  const requiredSpecialty = nlpSpecialtyToColumn(nlpData?.required_specialty);
  if (requiredSpecialty) {
    const load = getSpecialistLoad(row, requiredSpecialty, row.hospital_name);
    if (load == null) {
      if (acuity <= 2) score += noSpecialistPenalty;
    } else {
      score += specialistLoadWeight * load;
    }
  }
  return score;
}

async function fetchEtasFromMatrix(originLat, originLon, hospitals) {
  const apiKey = typeof process !== "undefined" && process.env?.REACT_APP_GMAPS_API_KEY;
  if (!apiKey) return null; // caller will use CSV column

  const origin = `${originLat},${originLon}`;
  const destinations = hospitals.map((h) => `${h.latitude},${h.longitude}`).join("|");
  const url = `https://maps.googleapis.com/maps/api/distancematrix/json?origins=${encodeURIComponent(origin)}&destinations=${encodeURIComponent(destinations)}&departure_time=now&key=${apiKey}`;

  try {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), 10000);
    const res = await fetch(url, { signal: controller.signal });
    clearTimeout(id);
    const response = await res.json();
    const etas = {};
    if (response?.status === "OK" && response?.rows?.[0]) {
      const elements = response.rows[0].elements;
      hospitals.forEach((h, i) => {
        const elem = elements[i];
        if (elem?.status === "OK") {
          const duration = elem.duration_in_traffic ?? elem.duration;
          etas[h.hospital_name] = duration ? duration.value / 60 : 999;
        } else etas[h.hospital_name] = 999;
      });
    } else hospitals.forEach((h) => (etas[h.hospital_name] = 999));
    return etas;
  } catch {
    return null;
  }
}

/**
 * Get the best hospital. Loads dataset from HOSPITAL_DATASET_PATH, applies hard filters, then scores (lowest = best).
 * @param {Object} nlpExtractedData - e.g. { acuity_level, required_specialty }
 * @param {number} currentLat
 * @param {number} currentLon
 * @returns {Promise<{hospital_id, hospital_name, routing_score, eta_minutes, ...}|null>}
 */
export async function getOptimalHospital(nlpExtractedData, currentLat, currentLon) {
  const allRows = await loadHospitalsFromDataset();
  const filtered = applyHardFilters(allRows, nlpExtractedData);
  if (filtered.length === 0) return null;

  const etasFromApi = await fetchEtasFromMatrix(currentLat, currentLon, filtered);
  const useCsvTravel = etasFromApi == null;

  const scored = filtered.map((row) => {
    const etaMinutes = useCsvTravel ? num(row.ambulance_travel_time_min_sim) : (etasFromApi[row.hospital_name] ?? 999);
    const score = scoreHospital(row, etaMinutes, nlpExtractedData);
    return { row, score, etaMinutes };
  });
  scored.sort((a, b) => a.score - b.score);
  const best = scored[0];
  const r = best.row;

  const requiredSpecialty = nlpSpecialtyToColumn(nlpExtractedData?.required_specialty);
  const specialist_load = requiredSpecialty
    ? getSpecialistLoad(r, requiredSpecialty, r.hospital_name)
    : null;
  const specialist_ready = requiredSpecialty ? hasSpecialistFor(r, requiredSpecialty, r.hospital_name) : true;

  return {
    hospital_id: r.hospital_name,
    hospital_name: r.hospital_name,
    routing_score: Math.round(best.score * 100) / 100,
    eta_minutes: Math.round(best.etaMinutes * 10) / 10,
    latitude: num(r.latitude),
    longitude: num(r.longitude),
    available_ed_beds: num(r.available_ed_beds_sim),
    available_icu_beds: num(r.available_icu_beds_sim),
    er_wait_min: num(r.er_wait_min_sim),
    trauma_level: r.trauma_level,
    stroke_center_level: r.stroke_center_level,
    cardiac_cath_lab: r.cardiac_cath_lab,
    specialist_ready,
    specialist_load: specialist_load != null ? Math.round(specialist_load * 100) / 100 : null,
  };
}

/**
 * Record a patient sent to this hospital. Reduces ED/ICU in overlay and increments specialist_patients
 * for each required specialty so future routing sees higher specialist load.
 * @param {string} hospitalId - hospital_name from getOptimalHospital result
 * @param {Object} nlpData - e.g. { acuity_level, required_specialty } or { required_specialties }
 */
export function sendPatient(hospitalId, nlpData) {
  if (!hospitalId) return;
  const acuity = Number(nlpData?.acuity_level) ?? 3;
  sentPatientOverlay[hospitalId] = sentPatientOverlay[hospitalId] || {
    edBedsDelta: 0,
    icuBedsDelta: 0,
    specialistPatientsDelta: {},
  };
  sentPatientOverlay[hospitalId].edBedsDelta -= 1;
  if (acuity <= 2) sentPatientOverlay[hospitalId].icuBedsDelta -= 1;
  for (const specialty of getRequiredSpecialtiesFromNlp(nlpData)) {
    if (!specialty) continue;
    const delta = sentPatientOverlay[hospitalId].specialistPatientsDelta;
    delta[specialty] = (delta[specialty] ?? 0) + 1;
  }
}

File: Algo.py
"""
Hospital routing algorithm — loads hospital data from CSV, scores per data_explanation.
get_optimal_hospital(nlp_extracted_data, current_lat, current_lon) only; dataset path defined below.
"""
import os
import csv
import requests

# Which dataset to load (path relative to this file's directory or cwd).
HOSPITAL_DATASET_PATH = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    "austin_hospitals_demo - austin_hospitals_demo.csv",
)

# Overlay when send_patient() is called: bed deltas and specialist_patients increments per hospital.
_sent_patient_overlay = {}  # hospital_name -> {"ed_beds_delta", "icu_beds_delta", "specialist_patients_delta": { specialty: int }}

SPECIALTY_NAMES = ("Cardiology", "Trauma", "Neurology")


def _nlp_specialty_to_column(specialty):
    """Map NLP required_specialty to CSV specialty name for load calculation."""
    if not specialty or (specialty or "").lower() == "general":
        return None
    s = specialty.lower()
    if "cardiac" in s or "stemi" in s or "heart" in s:
        return "Cardiology"
    if "trauma" in s:
        return "Trauma"
    if "stroke" in s or "neuro" in s:
        return "Neurology"
    return None


def _get_required_specialties_from_nlp(nlp_data):
    specs = nlp_data.get("required_specialties")
    if isinstance(specs, (list, tuple)):
        return [s for s in (_nlp_specialty_to_column(x) for x in specs) if s]
    one = _nlp_specialty_to_column(nlp_data.get("required_specialty"))
    return [one] if one else []


def _get_specialist_load(row, specialty, hospital_name):
    """Specialist load = (csv specialist_patients + overlay) / specialists. None if no specialists."""
    if not specialty or specialty not in SPECIALTY_NAMES:
        return None
    col_doc = f"specialists_{specialty}"
    col_pat = f"specialist_patients_{specialty}"
    num_doctors = _num(row.get(col_doc))
    if num_doctors <= 0:
        return None
    csv_patients = _num(row.get(col_pat))
    overlay = _sent_patient_overlay.get(hospital_name, {}).get("specialist_patients_delta", {})
    patients = max(0, csv_patients + overlay.get(specialty, 0))
    return patients / num_doctors


def _has_specialist_for(row, specialty, hospital_name):
    return _get_specialist_load(row, specialty, hospital_name) is not None

_cached_hospitals = None


def _load_hospitals_from_dataset():
    global _cached_hospitals
    if _cached_hospitals is not None:
        return _cached_hospitals
    if not os.path.isfile(HOSPITAL_DATASET_PATH):
        raise FileNotFoundError(f"Hospital dataset not found: {HOSPITAL_DATASET_PATH}")
    with open(HOSPITAL_DATASET_PATH, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        _cached_hospitals = list(reader)
    return _cached_hospitals


def _num(v):
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def _has_trauma_capability(trauma_level):
    if not trauma_level or trauma_level == "N/A":
        return False
    return trauma_level.upper().startswith("I")  # I, II, III, IV


def _has_stroke_capability(stroke_level):
    if not stroke_level:
        return False
    u = stroke_level.lower()
    if "none" in u and "capable" not in u:
        return False
    return "primary" in u or "comprehensive" in u or "capable" in u


def _has_cardiac_capability(cardiac_cath_lab):
    return (cardiac_cath_lab or "").strip().lower() == "yes"


def _has_pediatric_capability(pediatric_specialty):
    if not pediatric_specialty:
        return False
    u = pediatric_specialty.lower()
    return "yes" in u or "limited" in u or "nicu" in u


def _specialty_requires_capability(specialty):
    if not specialty or (specialty or "").lower() == "general":
        return None
    s = specialty.lower()
    if "trauma" in s:
        return "trauma"
    if "stroke" in s:
        return "stroke"
    if "cardiac" in s or "stemi" in s or "heart" in s:
        return "cardiac"
    if "pediatric" in s or "peds" in s:
        return "pediatric"
    return None


def _hospital_has_required_capability(row, required_cap):
    if not required_cap:
        return True
    if required_cap == "trauma":
        return _has_trauma_capability(row.get("trauma_level"))
    if required_cap == "stroke":
        return _has_stroke_capability(row.get("stroke_center_level"))
    if required_cap == "cardiac":
        return _has_cardiac_capability(row.get("cardiac_cath_lab"))
    if required_cap == "pediatric":
        return _has_pediatric_capability(row.get("pediatric_specialty"))
    return True


def _apply_hard_filters(rows, nlp_data):
    acuity = int(nlp_data.get("acuity_level", 3))
    required_cap = _specialty_requires_capability(nlp_data.get("required_specialty"))
    is_critical = acuity <= 2
    out = []
    for row in rows:
        if (row.get("ed_diversion_sim") or "").strip().lower() == "yes":
            continue
        if not _hospital_has_required_capability(row, required_cap):
            continue
        if is_critical:
            icu_avail = _num(row.get("available_icu_beds_sim"))
            overlay = _sent_patient_overlay.get(row["hospital_name"], {})
            icu_avail += overlay.get("icu_beds_delta", 0)
            if icu_avail <= 0:
                continue
        out.append(row)
    return out


def _score_hospital(row, eta_minutes, nlp_data):
    """Lower score = better. Includes specialist load (patients/doctors) for required_specialty."""
    acuity = int(nlp_data.get("acuity_level", 3))
    travel_weight = 1.0
    wait_weight = 0.5
    capacity_weight = 2.0
    staff_weight = 0.3
    specialist_load_weight = 20
    no_specialist_penalty = 500

    er_wait = _num(row.get("er_wait_min_sim"))
    ed_beds_raw = _num(row.get("available_ed_beds_sim"))
    overlay = _sent_patient_overlay.get(row["hospital_name"], {})
    available_ed_beds = max(0, ed_beds_raw + overlay.get("ed_beds_delta", 0))
    physicians = _num(row.get("on_call_ed_physicians_sim"))
    icu_avail = max(0, _num(row.get("available_icu_beds_sim")) + overlay.get("icu_beds_delta", 0))

    score = travel_weight * eta_minutes + wait_weight * er_wait
    score -= capacity_weight * available_ed_beds
    score -= staff_weight * physicians
    if acuity <= 2 and icu_avail <= 1:
        score += 50

    required_specialty = _nlp_specialty_to_column(nlp_data.get("required_specialty"))
    if required_specialty:
        load = _get_specialist_load(row, required_specialty, row["hospital_name"])
        if load is None:
            if acuity <= 2:
                score += no_specialist_penalty
        else:
            score += specialist_load_weight * load
    return score


def fetch_etas_from_matrix(origin_lat, origin_lon, rows):
    """One call to Google Distance Matrix; returns dict hospital_name -> eta minutes."""
    api_key = os.getenv("GMAPS_API_KEY")
    if not api_key:
        return None
    origin = f"{origin_lat},{origin_lon}"
    destinations = "|".join([f"{r['latitude']},{r['longitude']}" for r in rows])
    url = (
        "https://maps.googleapis.com/maps/api/distancematrix/json"
        f"?origins={origin}&destinations={destinations}&departure_time=now&key={api_key}"
    )
    try:
        response = requests.get(url, timeout=10).json()
    except (requests.RequestException, ValueError):
        return None
    etas = {}
    if response.get("status") == "OK" and response.get("rows"):
        elements = response["rows"][0]["elements"]
        for i, row in enumerate(rows):
            name = row["hospital_name"]
            if i < len(elements) and elements[i].get("status") == "OK":
                d = elements[i].get("duration_in_traffic") or elements[i].get("duration")
                etas[name] = (d["value"] / 60) if d else 999
            else:
                etas[name] = 999
    else:
        for row in rows:
            etas[row["hospital_name"]] = 999
    return etas


def get_optimal_hospital(nlp_extracted_data, current_lat, current_lon):
    """
    Load dataset from HOSPITAL_DATASET_PATH, apply hard filters, score (lowest = best).
    Returns best hospital dict or None. No hospital list argument — data comes from CSV.
    """
    all_rows = _load_hospitals_from_dataset()
    filtered = _apply_hard_filters(all_rows, nlp_extracted_data)
    if not filtered:
        return None

    etas_from_api = fetch_etas_from_matrix(current_lat, current_lon, filtered)
    use_csv_travel = etas_from_api is None

    scored = []
    for row in filtered:
        if use_csv_travel:
            eta_min = _num(row.get("ambulance_travel_time_min_sim"))
        else:
            eta_min = etas_from_api.get(row["hospital_name"], 999)
        score = _score_hospital(row, eta_min, nlp_extracted_data)
        scored.append((score, row, eta_min))
    scored.sort(key=lambda x: x[0])
    best_score, best_row, best_eta = scored[0]

    required_specialty = _nlp_specialty_to_column(nlp_extracted_data.get("required_specialty"))
    specialist_load = (
        _get_specialist_load(best_row, required_specialty, best_row["hospital_name"])
        if required_specialty
        else None
    )
    specialist_ready = (
        _has_specialist_for(best_row, required_specialty, best_row["hospital_name"])
        if required_specialty
        else True
    )

    return {
        "hospital_id": best_row["hospital_name"],
        "hospital_name": best_row["hospital_name"],
        "routing_score": round(best_score, 2),
        "eta_minutes": round(best_eta, 1),
        "latitude": _num(best_row.get("latitude")),
        "longitude": _num(best_row.get("longitude")),
        "available_ed_beds": _num(best_row.get("available_ed_beds_sim")),
        "available_icu_beds": _num(best_row.get("available_icu_beds_sim")),
        "er_wait_min": _num(best_row.get("er_wait_min_sim")),
        "trauma_level": best_row.get("trauma_level"),
        "stroke_center_level": best_row.get("stroke_center_level"),
        "cardiac_cath_lab": best_row.get("cardiac_cath_lab"),
        "specialist_ready": specialist_ready,
        "specialist_load": round(specialist_load, 2) if specialist_load is not None else None,
    }


def send_patient(hospital_id, nlp_data):
    """
    Record a patient sent to this hospital. Reduces ED/ICU in overlay and increments
    specialist_patients for each required specialty so future routing sees higher specialist load.
    """
    if not hospital_id:
        return
    _sent_patient_overlay.setdefault(
        hospital_id,
        {"ed_beds_delta": 0, "icu_beds_delta": 0, "specialist_patients_delta": {}},
    )
    _sent_patient_overlay[hospital_id]["ed_beds_delta"] -= 1
    acuity = int(nlp_data.get("acuity_level", 3))
    if acuity <= 2:
        _sent_patient_overlay[hospital_id]["icu_beds_delta"] -= 1
    delta = _sent_patient_overlay[hospital_id]["specialist_patients_delta"]
    for specialty in _get_required_specialties_from_nlp(nlp_data):
        delta[specialty] = delta.get(specialty, 0) + 1
